galera-autofix: Race condition kills cluster when autofix kicks in during a restore

**Describe the bug**
If galera-autofix kicks in while the cluster is performing an SST, it kills the current primary mysqld process to elect it in --force-new-cluster but the other nodes may not be killed to trash the state which results in the --force-new-cluster node to attempt to connect to the existing cluster that was created prior to it being restarted

**To Reproduce**
1. Stop cluster in order: 3, 2, 1
2. Start node 1 (minute:0)
4. Start node 2 and 3 at minute 8
5. Node 1, 2 and 3 are creating their cluster without the help of galera-autofix at minute 9
6. Galera autofix kicks in at minute 10
7. Galera autofix determines node 1 has the lastest data and kills mysqld and starts it in force-new-cluster
8. Nodes 2 and 3 haven't been killed yet because they didn't reach minute 10 so they stay alive and broadcast the old state of the cluster so the cluster becomes partitioned
9. Node 1 starts up but stays in NON-PRMARY which means the DB is started but is not healthy so its unable to sync its state to the rest of the cluster
10. On top of the DB being unhealthy the seqno reporting gets broken on node 1 because it doesn't read the live seqno which means the DB is stuck broken forever until a manual intervention is made

This issue will also happen if mysqld falls into NON-PRIMARY without dying

Example of how it can fall into that state:
```
Sep 17 10:51:50 pf mysqld: }2021-09-17 10:51:50 140339960526592 [ERROR] WSREP: exception from gcomm, backend must be restarted: node_uuid != uuid():  (FATAL)
Sep 17 10:51:50 pf mysqld:    at gcomm/src/evs_proto.cpp:set_inactive():1119
```

The fixes already described above will handle this case too

--- a/go/cmd/galera-autofix/main.go
+++ b/go/cmd/galera-autofix/main.go
@@ -90,8 +90,8 @@ func seqnoReporting(ctx context.Context) {
 	servers := pfconfigdriver.AllClusterServers{}
 	for {
 		seqno := mariadb.DefaultSeqno
-		if mariadb.IsLocalDBAvailable(ctx) {
-			seqno = getRecordLiveSeqno(ctx)
+		if liveSeqNo := mariadb.GetLiveSeqNo(ctx); liveSeqNo != DefaultSeqno {
+			seqno = liveSeqNo
 		} else {
 			var err error
 			seqno, err = mariadb.GetColdSeqno(ctx)

--- a/go/cmd/galera-autofix/main.go
+++ b/go/cmd/galera-autofix/main.go
@@ -90,7 +90,7 @@ func seqnoReporting(ctx context.Context) {
 	servers := pfconfigdriver.AllClusterServers{}
 	for {
 		seqno := mariadb.DefaultSeqno
-		if liveSeqNo := mariadb.GetLiveSeqNo(ctx); liveSeqNo != DefaultSeqno {
+		if liveSeqNo := mariadb.GetLocalLiveSeqno(ctx); liveSeqNo != mariadb.DefaultSeqno {
 			seqno = liveSeqNo
 		} else {
 			var err error

--- a/go/cmd/galera-autofix/main.go
+++ b/go/cmd/galera-autofix/main.go
@@ -87,14 +87,11 @@ func seqnoReporting(ctx context.Context) {
 		if liveSeqno := mariadb.GetLocalLiveSeqno(ctx); liveSeqno != mariadb.DefaultSeqno {
 			recordLiveSeqno(ctx, liveSeqno)
 			seqno = liveSeqno
+		} else if coldSeqno, err := mariadb.GetColdSeqno(ctx); err == nil {
+			seqno = coldSeqno
 		} else {
-			var err error
-			seqno, err = mariadb.GetColdSeqno(ctx)
-			if err != nil {
-				log.LoggerWContext(ctx).Error("Unable to obtain sequence number")
-				time.Sleep(seqnoReportingInterval)
-				continue
-			}
+			log.LoggerWContext(ctx).Warn("This server doesn't have a seqno. Will report the inexistant seqno. This node will have no chance to be the one running the latest data.")
+			seqno = mariadb.InexistantSeqno
 		}
 
 		pfconfigdriver.FetchDecodeSocketCache(ctx, &servers)

--- a/go/cmd/galera-autofix/main.go
+++ b/go/cmd/galera-autofix/main.go
@@ -74,24 +74,19 @@ func getSeqnoReport(ctx context.Context, nodes *NodeList) bool {
 	}
 }
 
-func getRecordLiveSeqno(ctx context.Context) int {
-	seqno := mariadb.GetLocalLiveSeqno(ctx)
-	if seqno != mariadb.DefaultSeqno {
-		log.LoggerWContext(ctx).Debug(fmt.Sprintf("Found the following live sequence number: %d", seqno))
-		err := ioutil.WriteFile(mariadb.GaleraAutofixSeqnoFile, []byte(fmt.Sprintf("%d", seqno)), 0644)
-		sharedutils.CheckError(err)
-	} else {
-		log.LoggerWContext(ctx).Debug("Failed to obtain the live sequence number")
-	}
-	return seqno
+func recordLiveSeqno(ctx context.Context, seqno int) {
+	log.LoggerWContext(ctx).Debug(fmt.Sprintf("Found the following live sequence number: %d", seqno))
+	err := ioutil.WriteFile(mariadb.GaleraAutofixSeqnoFile, []byte(fmt.Sprintf("%d", seqno)), 0644)
+	sharedutils.CheckError(err)
 }
 
 func seqnoReporting(ctx context.Context) {
 	servers := pfconfigdriver.AllClusterServers{}
 	for {
 		seqno := mariadb.DefaultSeqno
-		if liveSeqNo := mariadb.GetLocalLiveSeqno(ctx); liveSeqNo != mariadb.DefaultSeqno {
-			seqno = liveSeqNo
+		if liveSeqno := mariadb.GetLocalLiveSeqno(ctx); liveSeqno != mariadb.DefaultSeqno {
+			recordLiveSeqno(ctx, liveSeqno)
+			seqno = liveSeqno
 		} else {
 			var err error
 			seqno, err = mariadb.GetColdSeqno(ctx)

--- a/go/cmd/galera-autofix/main.go
+++ b/go/cmd/galera-autofix/main.go
@@ -265,7 +265,9 @@ func bootAndRejoinCluster(ctx context.Context, node *Node) {
 }
 
 func bootNewCluster(ctx context.Context) {
-	if mariadb.IsActive(ctx) {
+	if mariadb.IsLocalDBAvailable(ctx) {
+		log.LoggerWContext(ctx).Warn("The DB is now available, not booting a new cluster from this node")
+	} else if mariadb.IsActive(ctx) {
 		mariadb.ForceStop(ctx)
 		mariadb.StartNewCluster(ctx)
 	} else {

--- a/go/galeraautofix/mariadb/mariadb.go
+++ b/go/galeraautofix/mariadb/mariadb.go
@@ -18,6 +18,7 @@ import (
 
 const RunningSeqno = -1
 const DefaultSeqno = -2
+const InexistantSeqno = -3
 
 const GaleraAutofixSeqnoFile = "/var/lib/mysql/galera-autofix-seqno.dat"
 

